src/mesh/mod.rs

```diff
@@ -0,0 +1,234 @@
+mod column_builder;
+mod plane_builder;
+
+use std::ops::Add;
+
+pub use column_builder::ColumnMeshBuilder;
+pub use plane_builder::PlaneMeshBuilder;
+
+use glam::{Quat, Vec2, Vec3};
+
+use crate::{Hex, HexLayout};
+
+pub(crate) const BASE_FACING: Vec3 = Vec3::Y;
+pub(crate) const UV_DELTA: Vec2 = Vec2::splat(0.5);
+
+#[derive(Debug, Clone, Default)]
+/// Mesh information. The `const LEN` attribute ensures that there is the same number of vertices, normals and uvs
+pub struct MeshInfo {
+    /// All vertices positions information (`Vertex_Position` attribute)
+    pub vertices: Vec<Vec3>,
+    /// Normals for each vertex (You might need to swap `y` and `z`) (`Vertex_Normal` attribute)
+    pub normals: Vec<Vec3>,
+    /// UV coordinates of each vertex (`Vertex_Uv` attribute)
+    pub uvs: Vec<Vec2>,
+    /// Vertex indices for triangles
+    pub indices: Vec<u16>,
+}
+
+impl MeshInfo {
+    /// Returns a new [`MeshInfo`] but with vertex positions and normals rotated
+    #[inline]
+    #[must_use]
+    pub fn rotated(self, rotation: Quat) -> Self {
+        Self {
+            vertices: self
+                .vertices
+                .into_iter()
+                .map(|v| rotation.mul_vec3(v))
+                .collect(),
+            normals: self
+                .normals
+                .into_iter()
+                .map(|n| rotation.mul_vec3(n))
+                .collect(),
+            ..self
+        }
+    }
+
+    /// Returns a new [`MeshInfo`] but with `offset` applied to vertex positions
+    #[inline]
+    #[must_use]
+    pub fn with_offset(self, offset: Vec3) -> Self {
+        Self {
+            vertices: self.vertices.into_iter().map(|p| p + offset).collect(),
+            ..self
+        }
+    }
+
+    /// Merges `rhs` into `self`
+    ///
+    /// # Panics
+    ///
+    /// Will panic if there are more vertices than [`u16::MAX`]
+    pub fn merge_with(&mut self, rhs: Self) {
+        let indices_offset =
+            u16::try_from(self.vertices.len()).expect("MeshInfo has too many vertices");
+        self.vertices.extend(rhs.vertices);
+        self.normals.extend(rhs.normals);
+        self.uvs.extend(rhs.uvs);
+        self.indices
```

Comment:
It's surprising to me that there's no attempt at vertex deduplication being done here at all. Maybe the right choice for this method, but it should be documented.

Reply:
That would be much more complex as both vertex and index data would need to be edite, triangles should be regenerated, etc. That would be part of a bigger chunking PR

Reply:
Makes sense!